src/value/de.rs

```diff
@@ -0,0 +1,415 @@
+        }
+    }
+
+    fn deserialize_enum<T>(
+        self,
+        _name: &'static str,
+        _variants: &'static [&'static str],
+        visitor: T,
+    ) -> Result<T::Value, Error>
+    where
+        T: Visitor<'i>,
+    {
+        match self.value.rb_type() {
+            ruby_value_type::RUBY_T_STRING => {
+                let variant = unsafe { RString::from_rb_value_unchecked(self.value.as_rb_value()) };
+
+                visitor.visit_enum(EnumDeserializer {
+                    variant: variant.to_string()?,
+                    value: None,
+                })
+            }
+
+            ruby_value_type::RUBY_T_HASH => {
+                let hash = unsafe { RHash::from_rb_value_unchecked(self.value.as_rb_value()) };
+
+                if hash.len() == 1 {
+                    let key: String = hash.keys()?.entry(0)?;
+                    let value = hash.get(key.as_str());
+
+                    visitor.visit_enum(EnumDeserializer {
+                        variant: key,
+                        value,
+                    })
+                } else {
+                    Err(Error::new(
+                        crate::exception::type_error(),
+                        format!("can't deserialize Hash of length {} to Enum", hash.len()),
+                    ))
+                }
+            }
+
+            _ => Err(Error::new(
+                crate::exception::type_error(),
+                format!(
+                    "can't deserialize {} to Enum",
+                    unsafe { self.value.classname() }.into_owned()
+                ),
+            )),
+        }
+    }
+
+    fn deserialize_newtype_struct<T>(
+        self,
+        _name: &'static str,
+        visitor: T,
+    ) -> Result<T::Value, Error>
+    where
+        T: Visitor<'i>,
+    {
+        visitor.visit_newtype_struct(self)
+    }
+
+    fn deserialize_ignored_any<T>(self, visitor: T) -> Result<T::Value, Error>
+    where
+        T: Visitor<'i>,
+    {
+        visitor.visit_unit()
+    }
+
+    forward_to_deserialize_any! {
+        <T: Visitor<'i>>
+        bool i8 i16 i32 i64 i128 u8 u16 u32 u64 u128 f32 f64 char str string
+        unit unit_struct seq tuple tuple_struct map struct identifier
+    }
+}
+
+struct ArrayDeserializer<'i> {
+    array: &'i RArray,
+    index: isize,
+}
+
+impl<'i> ArrayDeserializer<'i> {
+    fn new(array: &'i RArray) -> Self {
+        Self { array, index: 0 }
+    }
+}
+
+impl<'i, 'de> SeqAccess<'de> for ArrayDeserializer<'i> {
+    type Error = Error;
+
+    fn next_element_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>, Error>
+    where
+        T: DeserializeSeed<'de>,
+    {
+        if self.index as usize >= self.array.len() {
+            return Ok(None);
+        }
+
+        let value: Value = self.array.entry(self.index)?;
+        self.index += 1;
+
+        seed.deserialize(Deserializer { value }).map(Some)
+    }
+}
+
+struct HashDeserializer<'i> {
+    hash: &'i RHash,
```

Comment:
I might be missing something here, but I don't think you need the reference here. All Ruby types (`RHash`, `Value`, etc) are basically pointers, so effectively they already are references (with the real data being owned by Ruby), and they all implement `Copy`, so like

``` rust
let a = RHash::new();
let b = a;
a.aset("foo", "bar").unwrap();
let x: String = b.fetch("foo").unwrap();
assert_eq!(x, "bar");
```